src/os2datascanner/utils/optimiser/optimiser.py

```diff
@@ -0,0 +1,314 @@
+                    future.extend(list(self.containify(k)))
+
+            self.containers.extend(self.get_containers(main))
+            self.containers = list(remove_duplicates(self.containers))
+
+            for cont in self.containers:
+                if isinstance(cont, CustomContainer):
+
+                    unique_components = list(remove_duplicates(cont.components.copy()))
+                    if not list_equality_no_order(cont.components.copy(), unique_components):
+                        cont.components = unique_components
+                        self.found_redundancy = True
+
+                    duplicates = list(find_duplicates(cont.components, cont.parent.components))
+
+                    if duplicates and cont != main:
+                        self.found_redundancy = True
+
+                        duplicates = list(find_duplicates(cont.components, cont.parent.components))
+
+                        if cont.operator == "and" and cont.parent.operator == "or":
+                            cont.parent.components = inverse_extend(cont.parent.components, duplicates)
+                        else:
+                            cont.components = inverse_extend(cont.components, duplicates)
+
+                        # On redundancy, always remove duplicate element from child container
+                        # except in the case where parent is OR and child is AND.
+
+                        # a OR b OR (b AND c) != a OR b OR (c AND True) [-]
+                        # a OR b OR (b AND c) == a OR (b AND c) [+]
+
+                    if cont.operator == cont.parent.operator and cont != main:
+                        try:
+                            cont.parent.components.remove(cont)
+                            self.containers.remove(cont)
+                            cont.parent.components.extend(cont.components)
+                            continue
+                        except ValueError:
+                            pass
+                    if self.check_empty(cont) or self.check_useless(cont) or self.check_symbol_redundancy(cont):
+                        continue
+
+        dump_json(main.as_dict(), self.OUT_PATH)
+        return self.cycles > 1 # If it did more than 1 cycle, it means it found something
+
+    def reformat_rule(self, path):
+        obj = str(load_json(path))
+        for id in rule_ids:
+            obj = obj.replace(str(id), rule_ids[id])
+        obj = obj.replace("'", '"') # Fixing some double | single quote problems
+        obj = json.loads(obj)
+        with open(OUT_PATH, "wt") as file:
+            json.dump(obj, file)
+
+    def run_optimiser(self):
+        refactored = True
+        while refactored:
+            refactored = self.optimisation_cycle()
+            # Get previous cycles. 
+            # While previous optimisation_cycle() found stuff, run it again.
+            # Maybe check for risk of infinite recursion
+        self.reformat_rule(OUT_PATH)
+
+
+class CustomContainer():
```

Comment:
Why are you doing your own JSON parsing when the `os2datascanner.engine2.rules` package exists? :thinking:

Reply:
I didnt know about the rules package when doing this. Also, did you mention the wrong piece of code or am i missing something in the 4 quoted lines ?